File: src/telemetry/login_data.cpp
#include "telemetry/login_data.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Telemetry
{
LoginData::LoginData(const PbTelemetry::LoginData& protobuf) {
    username = protobuf.username();
    password = protobuf.password();
}

LoginData::operator PbTelemetry::LoginData() const {
    PbTelemetry::LoginData ret;
    ret.set_username(username);
    ret.set_password(password);
    return ret;
}

std::string LoginData::serializeAsJsonString() const {
    PbTelemetry::LoginData protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string LoginData::serializeAsProtobufString() const {
    PbTelemetry::LoginData protobuf(*this);
    return protobuf.SerializeAsString();
}

bool LoginData::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::LoginData protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool LoginData::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::LoginData protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/telemetry/login_data.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict, Optional

from telemetry import login_data_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class LoginData:
    username: str = ""
    password: str = ""
    
    _proto_message: login_data_pb2.LoginData = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = login_data_pb2.LoginData()

    def _populate_proto(self):
        self._proto_message.username = self.username
        self._proto_message.password = self.password

    @classmethod
    def _from_proto(cls, proto_message) -> "LoginData":
        return cls(
            username = proto_message.username,
            password = proto_message.password,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "LoginData":
        message = login_data_pb2.LoginData()
        message.ParseFromString(data)
        return cls._from_proto(message)

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message, preserving_proto_field_name=True)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "LoginData":
        message = login_data_pb2.LoginData()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
